MdeModulePkg/Core/Dxe/Misc/MemoryProtectionSupport.c

```diff
@@ -255,6 +262,284 @@ SetUefiImageMemoryAttributes (
+  OUT UINTN                             *Count
+  )
+{
+  MEMORY_PROTECTION_SPECIAL_REGION_LIST_ENTRY  *SpecialRegionEntry;
+  LIST_ENTRY                                   *SpecialRegionEntryLink;
+  UINTN                                        Index = 0;
+
+  if ((SpecialRegions == NULL) || (*SpecialRegions != NULL)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if (mSpecialMemoryRegionsPrivate.Count == 0) {
+    return EFI_SUCCESS;
+  }
+
+  SortMemoryProtectionSpecialRegionList (&mSpecialMemoryRegionsPrivate.SpecialRegionList);
+
+  *SpecialRegions = AllocatePool (sizeof (MEMORY_PROTECTION_SPECIAL_REGION) * mSpecialMemoryRegionsPrivate.Count);
+
+  if (*SpecialRegions == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  *Count = mSpecialMemoryRegionsPrivate.Count;
+
+  for (SpecialRegionEntryLink = mSpecialMemoryRegionsPrivate.SpecialRegionList.ForwardLink;
+       SpecialRegionEntryLink != &mSpecialMemoryRegionsPrivate.SpecialRegionList;
+       SpecialRegionEntryLink = SpecialRegionEntryLink->ForwardLink)
+  {
+    SpecialRegionEntry = CR (
+                           SpecialRegionEntryLink,
+                           MEMORY_PROTECTION_SPECIAL_REGION_LIST_ENTRY,
+                           Link,
+                           MEMORY_PROTECTION_SPECIAL_REGION_LIST_ENTRY_SIGNATURE
+                           );
+
+    CopyMem (
+      &((*SpecialRegions)[Index++]),
+      &SpecialRegionEntry->SpecialRegion,
+      sizeof (MEMORY_PROTECTION_SPECIAL_REGION)
+      );
+  }
+
+  *Count = (Index == mSpecialMemoryRegionsPrivate.Count) ? mSpecialMemoryRegionsPrivate.Count : Index;
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Add the input MEMORY_PROTECTION_SPECIAL_REGION to the internal list of memory protection special regions
+  which will have the specified attributes applied when memory protections are initialized.
+
+  @param[in]  Start               Start of region which will be added to the special memory regions.
+                                  NOTE: This address will be page-aligned
+  @param[in]  Length              Length of the region which will be added to the special memory regions
+                                  NOTE: This value will be page-aligned
+  @param[in]  Attributes          Attributes to apply to the region during memory protection initialization
+
+  @retval   EFI_SUCCESS           SpecialRegion was successfully added
+  @retval   EFI_INVALID_PARAMTER  Length is zero
+  @retval   EFI_UNSUPPORTED       Memory protection has already been initialized
+  @retval   EFI_OUT_OF_RESOURCES  Failed to allocate memory
+**/
+EFI_STATUS
+EFIAPI
+AddSpecialRegion (
+  IN EFI_PHYSICAL_ADDRESS  Start,
+  IN EFI_PHYSICAL_ADDRESS  Length,
+  IN UINT64                Attributes
+  )
+{
+  MEMORY_PROTECTION_SPECIAL_REGION_LIST_ENTRY  *SpecialRegionEntry = NULL;
+  EFI_PHYSICAL_ADDRESS                         AlignedStart        = 0;
+  EFI_PHYSICAL_ADDRESS                         AlignedLength       = 0;

```

Comment:
If this is a length, even though EFI_PHYSICAL_ADDRESS resolves to a UINT64, it would be better to call this a UINT64.

Reply:
The length field is now a UINT64